packages/contractkit/src/identity/offchain-data-wrapper.ts

```diff
@@ -74,46 +89,81 @@ export default class OffchainDataWrapper {
 
   readDataFrom = makeAsyncThrowable(this.readDataFromAsResult.bind(this))
 
-  async writeDataTo(data: string, dataPath: string) {
+  async writeDataTo(
+    data: Buffer,
+    signature: string,
+    dataPath: string
+  ): Promise<OffchainErrors | void> {
     if (this.storageWriter === undefined) {
-      throw new Error('no storage writer')
+      return new NoStorageProvider()
+    }
+
+    try {
+      await Promise.all([
+        this.storageWriter.write(data, dataPath),
+        await this.storageWriter.write(
+          Buffer.from(trimLeading0x(signature), 'hex'),
+          dataPath + '.signature'
+        ),
+      ])
+    } catch (e) {
+      return new FetchError()
     }
-    await this.storageWriter.write(data, dataPath)
-    // TODO: Prefix signing abstraction
-    const sig = await NativeSigner(this.kit.web3.eth.sign, this.self).sign(data)
-    await this.storageWriter.write(sig, dataPath + '.signature')
   }
 }
 
 class StorageRoot {
-  constructor(readonly account: string, readonly address: string) {}
+  constructor(
+    readonly wrapper: OffchainDataWrapper,
+    readonly account: string,
+    readonly address: string
+  ) {}
+
+  async readAndVerifySignature(
+    dataPath: string,
+    buildTypedData: (x: Buffer) => Promise<EIP712TypedData>
+  ): Promise<Result<Buffer, OffchainErrors>> {
+    let dataResponse, signatureResponse
+
+    try {
+      ;[dataResponse, signatureResponse] = await Promise.all([
+        fetch(this.address + dataPath),
+        fetch(this.address + dataPath + '.signature'),
+      ])
+    } catch (error) {
+      return Err(new FetchError())
+    }
 
-  // TODO: Add decryption metadata (i.e. indicates ciphertext to be decrypted/which key to use)
-  async readAsResult(dataPath: string): Promise<Result<string, OffchainErrors>> {
-    const data = await fetch(this.address + dataPath)
-    if (!data.ok) {
+    if (!dataResponse.ok) {
       return Err(new FetchError())
     }
+    if (!signatureResponse.ok) {
+      return Err(new InvalidSignature())
+    }
 
-    const body = await data.text()
+    const [dataBody, signatureBody] = await Promise.all([
+      dataResponse.arrayBuffer(),
+      signatureResponse.arrayBuffer(),
+    ])
+    const body = Buffer.from(dataBody)
+    const signature = ensureLeading0x(Buffer.from(signatureBody).toString('hex'))
 
-    const signature = await fetch(this.address + dataPath + '.signature')
+    const typedData = await buildTypedData(body)
+    const guessedSigner = guessEIP712TypedDataSigner(typedData, signature)
 
-    if (!signature.ok) {
-      return Err(new InvalidSignature())
+    if (eqAddress(guessedSigner, this.account)) {
+      return Ok(body)
     }
 
-    // TODO: Compare against registered on-chain signers or off-chain signers
```

Comment:
Remembering this TODO, I believe we will still need this if the account is a smart contract (since it cannot sign itself, not create an `AuthorizedSigner` payload

Reply:
I'm not totally sure what to do. How do we check if it's a smart contract?

Reply:
Calling the `getCode` RPC is the simplest method. https://ethereum.stackexchange.com/a/28527

Reply:
I wouldn't check if its a contract or not, I would just check if the account is registered on `Accounts.sol` and if so, check against the authorized signers, similarly to how we do it in metadata: https://github.com/celo-org/celo-monorepo/blob/master/packages/contractkit/src/identity/metadata.ts#L56